[Breaking] New Custom Application docs for v21 release. (#2431)
commercetools/merchant-center-application-kit: website/src/content/migrating-from-project-level-custom-applications.mdx

---
title: Migrating from Project-level Custom Applications
---

When we released Custom Applications a couple of years ago we had one main goal: to provide a **great developer experience** for developing and extending Merchant Center functionalities.
Over time we have constantly improved this experience and broadened our set of tools and components to further help developers focussing on what matters: productivity.

We also gathered lots of valuable feedback both from our internal teams and especially from external partners and developers.

As a result, we started working on new important features and improvements of all aspects of Custom Applications, from documentation to database models.
In this document we explain what's new, what changed, and how to actually migrate.

<Warning>

Please take your time and read carefully this document to familiarize yourself with the changes, new concepts, and steps required to migrate your existing Custom Applications.

**Project-level Custom Applications must be migrated until the end of July 2022**.<br />

> During this period, Project-level Custom Applications still work but are considered in **maintenance mode**. **Adding new Project-level Custom Applications won't be possible anymore.**

Author: Good point, I'll mention that they won't be available anymore.

</Warning>

# Introducing (Org-level) Custom Applications

The new release of Custom Applications is what we call and refer to as **Org-level Custom Applications**.

Before we go into details about the new release and features, let's define some of the terminology used in the documentation:

* **Org-level Custom Applications** refers to the new release and it's the default functionality from now on. Application Kit packages (`@commercetools-frontend/*`) must use version `>= 21`.
* **Project-level Custom Applications** refers to the "legacy" functionality up until now. This is still available in maintenance mode for the [migration period](#migration-period) of time and will be removed afterwards.

<Info>

The documentation for Project-level Custom Applications [remains available at a different location](https://docs.commercetools.com/custom-applications/_legacy]) for the [migration period](#migration-period).

</Info>

> Up until now, Project-level Custom Applications had to be defined on a per-project level (hence the term) under "Settings > Custom Applications." After registering a new Custom Application you could activate it and start using it.

Author: For legacy we were mostly using "registering" wording.

## Choose the `entryPointUriPath`

The `entryPointUriPath` plays a crucial role in a Custom Application. Effectively, it's the **identifier** of the Custom Application similarly to the `projectKey` of the commercetools platform.

<Info>

From now on, with the introduction of Org-level Custom Applications, the `entryPointUriPath` is **considered unique** within each [cloud region](https://docs.commercetools.com/api/general-concepts#regions) environment, meaning that there can't be any duplicates.

</Info>

Choosing the `entryPointUriPath` becomes then more important and should be done carefully as it affects routing and permissions.

The `entryPointUriPath` value must adhere to the following restrictions:

* Only lowercase alphanumeric characters are allowed.
* Must be between 2 and 64 characters.
* Underscores and hyphens are allowed except as leading, trailing, and adjacent characters.

Furthermore, the `entryPointUriPath` value is tightly bound to User Permissions of the Custom Application and is used to derive the unique name of the User Permissions.<br />
The name consists of the following format: `{View,Manage}<EntryPointUriPath>`.

> Here are some examples:

Author: Yes correct

Over time we have constantly improved this experience and broadened our set of tools and components to further help developers focusing on what matters: productivity.

We also gathered lots of valuable feedback both from our internal teams and especially from external partners and developers.

As a result, we started working on new important features and improvements of all aspects of Custom Applications, from documentation to database models.
In this document we explain what's new, what changed, and how to actually migrate.

<Warning>

Please take your time and read this document carefully to familiarize yourself with the changes, new concepts, and steps required to migrate your existing Custom Applications.

**Project-level Custom Applications must be migrated until the end of July 2022**.<br />
During this period, Project-level Custom Applications still work but are considered in **maintenance mode**. **Adding new Project-level Custom Applications won't be possible anymore.**

After the migration deadline, Project-level Custom Applications won't be accessible anymore.

</Warning>

> # Introducing (Org-level) Custom Applications

Reviewer: ```suggestion
# Introducing Org-level Custom Applications
```

The [new release](/releases/2022-01-25-custom-applications-v21) of Custom Applications is what we call and refer to as **Org-level Custom Applications**.

Before we go into details about the new release and features, let's define some of the terminology used in the documentation:

* **Org-level Custom Applications** refers to the new release and it's the default functionality from now on. Application Kit packages (`@commercetools-frontend/*`) must use version `>= 21`.
* **Project-level Custom Applications** refers to the legacy functionality up until now. This is still available in maintenance mode during the [migration period](#migration-period) and will be removed afterwards.

<Info>

The documentation for Project-level Custom Applications [remains available](https://docs.commercetools.com/custom-applications/_legacy]) for the [migration period](#migration-period).

</Info>

Up until now, Project-level Custom Applications had to be defined on a per Project level (hence the term) under **Settings > Custom Applications**. After registering a new Custom Application you could activate it and start using it.

> However, the Custom Application was only available for that specific Project. If you wanted to use the Custom Application in another Project, you would need to go through the registration process all over again.

Reviewer: ```suggestion
The Custom Application was only available for that specific Project. If you wanted to use the Custom Application in another Project, you would need to go through the registration process all over again.
```

With the introduction of Org-level Custom Applications we intend to improve that process and enable new features for the time being and for the future.

What this means is that managing Custom Applications is done now at the Organization level rather than the Project level. One of the main advantages is being able to **configure a Custom Application once, and use it in multiple projects**.

# OAuth Scopes and User Permissions

> Up until now, Project-level Custom Applications weren't able to define which [OAuth Scopes](https://docs.commercetools.com/api/scopes) of the commercetools Platform were needed and could only rely on the "implicit" OAuth Scopes defined for User Permissions of the "native" Merchant Center applications (**Organizations > Teams > Permissions**).

Author: We usually refer to our internal applications as "internal" or "native".

> Up until now, Project-level Custom Applications weren't able to define which [OAuth Scopes](https://docs.commercetools.com/api/scopes) of the commercetools Platform were needed and could only rely on the "implicit" OAuth Scopes defined for User Permissions of the "native" Merchant Center applications (**Organizations > Teams > Permissions**).

Author: FYI: we settled to _default_.

> Let us look at an example. The User Permission `ViewProducts` maps internally to some OAuth Scopes, like `view_products`, `view_states`, `view_types`, etc. If a Custom Application had to use the [States API](https://docs.commercetools.com/api/projects/states), it would need to use the `ViewProducts` permission to properly authorize the User to perform the API requests.

Author: @adinakleine Can you explain when should I use capital letters (Products, Organizations, Users) and when should it be lowercase?

> Let us look at an example. The User Permission `ViewProducts` maps internally to some OAuth Scopes, like `view_products`, `view_states`, `view_types`, etc. If a Custom Application had to use the [States API](https://docs.commercetools.com/api/projects/states), it would need to use the `ViewProducts` permission to properly authorize the User to perform the API requests.

Reviewer: The list of commercetools terms helps here.

As a result, Custom Applications OAuth Scopes and User Permissions become completely **decoupled** from the "native" Merchant Center applications and allow to actually fulfill many more use cases for Custom Applications.

# Why should you migrate?

The Project-level Custom Applications will become end-of-life after the [migration period](#migration-period). After that, all related functionalities will be removed.

Migrating to Org-level Custom Applications not only brings new features and functionalities but it allows to benefit from any future improvements on top of that.

# Step by step

Migrating your existing Project-level Custom Applications to the new Org-level Custom Applications is fairly straight-forward but it requires first to understand the new concepts like [OAuth Scopes and User Permissions](#oauth-scopes-and-user-permissions).

As for the Custom Application code, nothing much needs to be changed except for the configuration file and the [permission names](#migrate-user-permissions) (in case they were used).

> In general there are 2 parts of the migration: development changes and configuration of the new Custom Application in the Merchant Center.

Author: The 2 parts refer to 1) changes in the codebase and 2) configuring/installing the Custom Application in the MC (Account > Orgs)

</Info>

Choosing the `entryPointUriPath` becomes then more important and should be done carefully as it affects routing and permissions.

The `entryPointUriPath` value must adhere to the following restrictions:

* Only lowercase alphanumeric characters are allowed.
* Must be between 2 and 64 characters.
* Underscores and hyphens are allowed except as leading, trailing, and adjacent characters.

Furthermore, the `entryPointUriPath` value is bound to User Permissions of the Custom Application and is used to derive the unique name of the User Permissions.<br />
The name consists of the following format: `{View,Manage}<EntryPointUriPath>`.

Here are some examples:

| `entryPointUriPath` | User Permission |
| --- | --- |
| `avengers` | `{View,Manage}Avengers` |
| `the-avengers` | `{View,Manage}TheAvengers` |
| `the_avengers` | `{View,Manage}The_Avengers` |
| `avengers-01` | `{View,Manage}Avengers/01` |
| `avengers_01` | `{View,Manage}Avengers_01` |

> 

Reviewer: I am wondering if we could use a brand free example. I am worried that we could get in trouble using "Avengers".

> This ensures that User Permissions are also unique for each Custom Application.

Reviewer: ```suggestion
This ensures that user permissions are unique for each Custom Application too.
```

> ## Migrate User Permissions

Reviewer: ```suggestion
## Migrate user permissions
```

> If the Custom Application was relying on existing User Permissions such as `ViewProducts`, `ManageOrders`, etc. you need to change that.

Reviewer: ```suggestion
If the Custom Application was relying on existing user permissions such as `ViewProducts` or `ManageOrders`, you need to change that.
```

> As explained in [OAuth Scopes and User Permissions](#oauth-scopes-and-user-permissions), Custom Applications are now decoupled from the "native" Merchant Center applications and thus from their User Permissions.

Reviewer: ```suggestion
As explained in [OAuth Scopes and user permissions](#oauth-scopes-and-user-permissions), Custom Applications are now decoupled from the "native" Merchant Center applications and thus from their user permissions.
```
Perhaps change "native" (see above)

In the Custom Application config you need to explicitly define the `oAuthScopes` needed by the Custom Application and map them in the `oAuthScopes.view` and `oAuthScopes.manage` fields.

> These fields relate to the unique User Permissions pair of your Custom Application, as explained in [Choose the `entryPointUriPath`](#choose-the-entrypointuripath), and are the only permission values that your Custom Application is able to use.

Reviewer: ```suggestion
These fields relate to the unique user permissions pair of your Custom Application, as explained in [Choose the `entryPointUriPath`](#choose-the-entrypointuripath), and are the only permission values that your Custom Application is able to use.
```

> In your Custom Application code, replace the usage of the "native" User Permissions with the name derived from the `entryPointUriPath`.

Reviewer: ```suggestion
In your Custom Application code, replace the "native" user permissions with the name derived from the `entryPointUriPath`.
```
Perhaps replace "native" (see above)

```js highlightLines="6"
import { useIsAuthorized } from '@commercetools-frontend/permissions';

const MyComponent = () => {
  const canViewAvengers = useIsAuthorized({
    // Assuming that the `entryPointUriPath` is `avengers`.
    demandedPermissions: ['ViewAvengers'],
  });
};
```

> Furthermore, you also need to replace the usage of the "native" User Permissions in the menu links `permissions` field in the Custom Application config.

Reviewer: ```suggestion
Furthermore, replace the "native" user permissions in the menu links `permissions` field in the Custom Application config.
```

```json highlightLines="6"
{
  "mainMenuLink": {
    "permissions": ["ViewAvengers"]
  }
}
```

## Remove `menu.json`

If your Custom Application uses the `menu.json` file, you need to migrate its content to the Custom Application config file. See [Update the Custom Application config](#update-the-custom-application-config) for more information.

In addition to that, you should also remove the `DEV_ONLY__loadNavbarMenuConfig` prop in the `<ApplicationShell>`.

## Development login via OIDC

<Info>

If you are already using this feature via the opt-in environment variable `ENABLE_OIDC_FOR_DEVELOPMENT=true`, you can remove the variable as this is now the default behavior.

</Info>

> To make use of the [OAuth Scopes and User Permissions](#oauth-scopes-and-user-permissions) in local development, even though the Custom Application hasn't been configured in the Merchant Center yet, the login workflow works a bit differently from before.

Reviewer: ```suggestion
To use the [OAuth Scopes and user permissions](#oauth-scopes-and-user-permissions) in local development, even though the Custom Application hasn't been configured in the Merchant Center yet, the login workflow works a bit differently from before.
```

The session is saved in the `sessionStorage` (not a cookie like in the production environments) and is sent with each API request as an HTTP `Authorization` header. It has a short expiration time (hours, not days) and is bound by default to the Project specified in the `env.development.initialProjectKey` and the OAuth Scopes specified in the `oAuthScopes` config field.

You can still switch between Projects and the session token is automatically replaced in the background without having to log in again.

<Warning>

Local development and login is only allowed for users being part of the `Administrators` Team of your Organization.

</Warning>

## Assign the application ID

Before deploying the Custom Application to production with the new setup, you need to define the `env.production.applicationId` in the Custom Application config.

<Info>

As long as you are developing the Custom Application locally, you can define any random value as it's a required field.

</Info>

> The Custom Application ID is provided when you add a Custom Application in the Merchant Center (**Organizations > Custom Applications > Configure > Add**). See [Managing Custom Applications in the Merchant Center](https://docs.commercetools.com/merchant-center/managing-custom-applications) for more information.

Reviewer: ```suggestion
The Custom Application ID is provided when you add a Custom Application to the Merchant Center (**Organizations > Custom Applications > Configure > Add**). See [Managing Custom Applications in the Merchant Center](https://docs.commercetools.com/merchant-center/managing-custom-applications) for more information.
```

## Use the new Custom Application in the Merchant Center

Depending on your deployment setup you might need to consider what migration strategy works best for you and the users of your Custom Application.

<Info>

> We recommend informing your company and colleagues that are using the Custom Application about the migration and possible unavailability of it during the time of the switch.

Reviewer: ```suggestion
We recommend informing the users in your organization about the upcoming migration and a possible unavailability of the Custom Applications during the migration period.
```

</Info>

> In general, before deploying the new changes, you should configure the Custom Application in the Merchant Center first (see [Managing Custom Applications in the Merchant Center](https://docs.commercetools.com/merchant-center/managing-custom-applications)) and assign the `applicationId` to the Custom Application config file.

Reviewer: ```suggestion
In general, before deploying the changes, you should configure the Custom Application in the Merchant Center first (see [Managing Custom Applications in the Merchant Center](https://docs.commercetools.com/merchant-center/managing-custom-applications)) and assign the `applicationId` to the Custom Application config file.
```

> At this point we recommend deactivating and removing the "old" Custom Application. You can then proceed with the deployment of the new setup.

Reviewer: ```suggestion
At this point we recommend deactivating and removing the deprecated Custom Application. You can then proceed with the deployment of the new setup.
```

> Finally, you need to install the Custom Application in your Organization and grant access to one or more projects.

Reviewer: ```suggestion
Finally, you need to install the Custom Application in your Organization and grant access to one or more Projects.
```

The Custom Application can now be used.

> ## Assign User Permissions to Teams

Reviewer: ```suggestion
## Assign user permissions to Teams
```

One of the new features of the Org-level Custom Applications is the ability to assign [OAuth Scopes and User Permissions](#oauth-scopes-and-user-permissions) to each Custom Application.

> Once the Custom Application has been installed to an Organization, admins of the Organization can [assign User Permissions to each Team](https://docs.commercetools.com/merchant-center/user-permissions).

Reviewer: ```suggestion
Once the Custom Application has been installed to an Organization, admins of the Organization can [assign user permissions to each Team](https://docs.commercetools.com/merchant-center/user-permissions).
```

<Info>

> As usual, User Permissions for the `Administrators` Team are automatically managed when Project access is granted or removed.

Reviewer: ```suggestion
User permissions for the `Administrators` Team are automatically managed when Project access is granted or removed.
```
